sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala

```diff
@@ -96,6 +98,31 @@ class CollationSQLRegexpSuite
     }
   }
 
+  test("RegExpReplace throws the right exception when replace fails on a particular row") {
+    val tableName = "regexpReplaceException"
+    withTable(tableName) {
+      Seq("NO_CODEGEN", "CODEGEN_ONLY").foreach { codegenMode =>
+        withSQLConf("spark.sql.codegen.factoryMode" -> codegenMode) {
+          sql(s"CREATE TABLE IF NOT EXISTS $tableName(s STRING)")
+          sql(s"INSERT INTO $tableName VALUES('first last')")
+          val query = s"SELECT regexp_replace(s, '(?<first>[a-zA-Z]+) (?<last>[a-zA-Z]+)', " +
+            s"'$$3 $$1') FROM $tableName"
+          val df = sql(query)
+          val plan = df.queryExecution.executedPlan
+          assert(plan.isInstanceOf[WholeStageCodegenExec] == (codegenMode == "CODEGEN_ONLY"))
+          val exception = intercept[SparkRuntimeException] {
+            df.collect()
+          }
+          assert(exception.getCondition == "INVALID_REGEXP_REPLACE")
+          assert(exception.getMessage.contains("""Could not perform regexp_replace for """ +
+            """`source = "first last"`, `pattern = "(?<first>[a-zA-Z]+) (?<last>[a-zA-Z]+)"`, """ +
+            """`replacement = "$3 $1"` and `position = 1`."""))
```

Comment:
Could you use `checkError`, please. We try to avoid embedding error messages to source code because of:
- tech editors and others should be able to edit error formats without repackaging Spark
- we might have error messages in other languages in the future.